Sources/NIO/Bootstrap.swift

```diff
@@ -352,6 +361,17 @@ public final class ClientBootstrap {
     ///
     /// The connected `Channel` will operate on `ByteBuffer` as inbound and `IOData` as outbound messages.
     ///
+    /// - warning: The `handler` closure may be invoked _multiple times_ so always instantiate
+    ///            `ChannelHandler`s within `handler`. The reason `handler` may be invoked multiple
+    ///            times is that to successfully set up a connection multiple connections might be
+    ///            setup in the process. Assuming a hostname that resolves to both IPv4 and IPv6
+    ///            addresses, NIO will follow [_Happy Eyeballs_](https://en.wikipedia.org/wiki/Happy_Eyeballs)
+    ///            and race both an IPv4 and an IPv6 connection. It is possible that both
+    ///            connections get fully established before the IPv4 connection will be closed
+    ///            again because the IPv6 connection 'won the race'. Therefore the
+    ///            `channelInitializer` might be called multiple times and it's important not to
+    ///            share stateful `ChannelHandler`s in more than one `Channel`.
```

Comment:
@weissi I think this is not a 100 % correct advice... Sometimes you want to share the handler even if it has state (for example if you would count the connections). I think we should just say that if its stateful and shared it must be thread-safe.

Reply:
@normanmaurer I know, found it hard to get the message across that "always instantiate inside, unless you know what you're doing :P"

Reply:
ok, have an idea, let's phrase like this:

> Therefore stateful `ChannelHandler`s should generally be instantiated within the closure. There are use-cases where a stateful `ChannelHandler` needs to be shared between multiple `Channel`s in which case the user is responsible for correctly synchronising the access to the state.